### PR TITLE
WIP: Pointclouds 

### PR DESCRIPTION
**Pull request by PeterPetrik**

### CODE REVIEW

File: python/CMakeLists.txt [hunk 1]
```diff
@@ -124,6 +124,7 @@ INCLUDE_DIRECTORIES(
   ${CMAKE_SOURCE_DIR}/src/core/processing
   ${CMAKE_SOURCE_DIR}/src/core/processing/models
   ${CMAKE_SOURCE_DIR}/src/core/providers
+  ${CMAKE_SOURCE_DIR}/src/core/pointcloud
```
Comment (nyalldawson) [inline python/CMakeLists.txt:127]: Don't forget to add this directory to `doc\CMakeLists.txt` too

File: python/CMakeLists.txt [hunk 2]
```diff
@@ -161,6 +162,7 @@ IF (WITH_GUI)
     ${CMAKE_SOURCE_DIR}/src/gui/locator
     ${CMAKE_SOURCE_DIR}/src/gui/numericformats
     ${CMAKE_SOURCE_DIR}/src/gui/processing
+    ${CMAKE_SOURCE_DIR}/src/gui/pointcloud
```
Comment (nyalldawson) [inline python/CMakeLists.txt:165]: Same with this

File: python/gui/auto_generated/qgsabstractdatasourcewidget.sip.in [hunk 3]
```diff
@@ -99,6 +99,13 @@ Emitted when a mesh layer has been selected for addition.
 Emitted when a vector tile layer has been selected for addition.
 
 .. versionadded:: 3.14
+%End
+
+    void addPointCloudLayer( const QString &url, const QString &baseName );
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsabstractdatasourcewidget.sip.in:104]: This should include the provider name too, just for future flexibility

File: src/app/qgisapp.cpp [hunk 4]
```diff
@@ -5746,6 +5758,48 @@ QgsVectorTileLayer *QgisApp::addVectorTileLayerPrivate( const QString &url, cons
   return layer.release();
 }
 
+QgsPointCloudLayer *QgisApp::addPointCloudLayerPrivate( const QString &uri, const QString &baseName, bool guiWarning )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:5761]: QgisApp::openLayer will also need to call addPointCloudLayerPrivate to allow direct drag and drop of point cloud sources to the canvas from file managers.

File: src/core/pointcloud/qgspointcloudprovidermetadata.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+
+#include "qgsprovidermetadata.h"
+
+///@cond PRIVATE
+#define SIP_NO_FILE
+
+/**
+ * This metadata class does not support creation of provider instances, because
+ * vector tile layer currently does not have a concept of data providers. This class
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudprovidermetadata.h:29]:
```suggestion
 * point cloud layer currently does not have a concept of data providers. This class
```

**Review by nyalldawson — COMMENTED** [review]
Looking over this one outstanding question I have is regarding a point cloud data provider. I really think we should plan for the future and seperate the provider out from the layer class, just like we do for rasters/vectors/mesh. Vector tiles were really an exception here, but even with those I was almost blocked from adding the support for ArcGIS vectortileservers because of the missing provider abstraction (it's just fortunate that this is one case were ESRI followed the standards close enough that I could work around this). (and I definitely think we'll see the need for non-pdal providers sometime in future)

**Comment by PeterPetrik** [comment]
> ... I really think we should plan for the future and seperate the provider out from the layer class, just like we do for rasters/vectors/mesh... 

@nyalldawson the issue here is we do not know *what* should be in the data provider yet... should it be access to raw data ? if that is the case, in the current branch we do not have such access at all... should it be access to EPT/indexed data? if that is the case, pdal cannot give you that... or some combination or raw and indexed?

**Comment by nyalldawson** [comment]
@PeterPetrik 
> the issue here is we do not know what should be in the data provider yet... should it be access to raw data ? if that is the case, in the current branch we do not have such access at all... should it be access to EPT/indexed data? if that is the case, pdal cannot give you that... or some combination or raw and indexed?

I would literally just make this a bare minimum shell at the moment, which implements ONLY the pure virtual methods coming from QgsDataProvider. The reason I'd like to see this done is that it ensures that the "wider" api is designed to accept a point cloud provider id -- so things like QgisApp::addPointCloudLayer will already be designed around having a provider id and we won't run into any possible issues with stable API in future if we do need to add a different point cloud provider type.

Does that make sense?

**Pull request closed without merging**